File: patterns/04-concurrency-patterns/04-fan-out-fan-in/python/fan_out_fan_in.py
class PatternError(Exception):
    def __init__(self, code): super().__init__(code); self.code=code
def evaluate(input_data):
    mode=input_data.get("mode"); tasks=input_data.get("tasks",[])
    if mode not in ("all","first_success"): raise PatternError("unknown_mode")
    if mode=="first_success" and input_data.get("side_effecting",False): raise PatternError("fanout_not_allowed")
    if len({task["id"] for task in tasks})!=len(tasks): raise PatternError("duplicate_task")
    ordered=sorted(enumerate(tasks),key=lambda pair:(pair[1]["complete_at"],pair[0])); completion=[];results=[];failures=[]
    for position,(_,task) in enumerate(ordered):
        completion.append(task["id"])
        if task["outcome"]=="success":
            results.append(f'{task["id"]}:{task["value"]}')
            if mode=="first_success": return {"completion_order":completion,"results":results,"failures":failures,"cancelled":[later[1]["id"] for later in ordered[position+1:]],"status":"completed"}
        elif task["outcome"]=="failure": failures.append(task["id"])
        else: raise PatternError("unknown_outcome")
    if mode=="first_success": raise PatternError("all_tasks_failed")
    return {"completion_order":completion,"results":results,"failures":failures,"cancelled":[],"status":"partial" if failures else "completed"}

File: patterns/04-concurrency-patterns/04-fan-out-fan-in/go/fan_out_fan_in.go
package fanoutfanin

import (
	"sort"
)

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type task struct {
	id, outcome, value string
	at, index          int
}

func strings(items []task) []string {
	out := []string{}
	for _, item := range items {
		out = append(out, item.id)
	}
	return out
}
func Evaluate(input map[string]any) (any, error) {
	mode, _ := input["mode"].(string)
	if mode != "all" && mode != "first_success" {
		return nil, &PatternError{code: "unknown_mode"}
	}
	if mode == "first_success" {
		if side, ok := input["side_effecting"].(bool); ok && side {
			return nil, &PatternError{code: "fanout_not_allowed"}
		}
	}
	raw, _ := input["tasks"].([]any)
	tasks := []task{}
	seen := map[string]bool{}
	for index, item := range raw {
		value := item.(map[string]any)
		id := value["id"].(string)
		if seen[id] {
			return nil, &PatternError{code: "duplicate_task"}
		}
		seen[id] = true
		tasks = append(tasks, task{id: id, outcome: value["outcome"].(string), value: value["value"].(string), at: int(value["complete_at"].(float64)), index: index})
	}
	sort.SliceStable(tasks, func(i, j int) bool { return tasks[i].at < tasks[j].at })
	completion := []string{}
	results := []string{}
	failures := []string{}
	for position, item := range tasks {
		completion = append(completion, item.id)
		switch item.outcome {
		case "success":
			results = append(results, item.id+":"+item.value)
			if mode == "first_success" {
				return map[string]any{"completion_order": completion, "results": results, "failures": failures, "cancelled": strings(tasks[position+1:]), "status": "completed"}, nil
			}
		case "failure":
			failures = append(failures, item.id)
		default:
			return nil, &PatternError{code: "unknown_outcome"}
		}
	}
	if mode == "first_success" {
		return nil, &PatternError{code: "all_tasks_failed"}
	}
	status := "completed"
	if len(failures) > 0 {
		status = "partial"
	}
	return map[string]any{"completion_order": completion, "results": results, "failures": failures, "cancelled": []string{}, "status": status}, nil
}
